pkg/tsdb/grafana-testdata-datasource/admission_handler.go

```diff
@@ -0,0 +1,87 @@
+package testdatasource
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ValidateAdmission implements backend.AdmissionHandler.
+func (s *Service) ValidateAdmission(ctx context.Context, req *backend.AdmissionRequest) (*backend.ValidationResponse, error) {
+	rsp, err := s.MutateAdmission(ctx, req)
+	if rsp != nil {
+		return &backend.ValidationResponse{
+			Allowed:  rsp.Allowed,
+			Result:   rsp.Result,
+			Warnings: rsp.Warnings,
+		}, err
+	}
+	return nil, err
+}
+
+// MutateAdmission implements backend.AdmissionHandler.
+func (s *Service) MutateAdmission(ctx context.Context, req *backend.AdmissionRequest) (*backend.MutationResponse, error) {
+	expected := (&backend.DataSourceInstanceSettings{}).GVK()
+	if req.Kind.Kind != expected.Kind && req.Kind.Group != expected.Group {
+		return getBadRequest("expected DataSourceInstanceSettings protobuf payload"), nil
+	}
+
+	// Convert the payload from protobuf to an SDK struct
+	settings, err := backend.DataSourceInstanceSettingsFromProto(req.ObjectBytes, "")
```

Comment:
Don't you need to provide the pluginID argument?

Reply:
not here -- since we don't copy it out anyway.  But I suppose would not hurt